Feature/leveldb

MoonSeongKwang opened the pull request:
chain과 chain_plugin의 구조 재설계와 levelDB의 world, eden chain 정보를 추가한 코드입니다.
kv_store.cpp의 saveXXX 함수들은 일단 기능만 확인 가능한 정도로만 짜두었으므로, 추후 수정할 예정입니다.

MoonSeongKwang committed:
Merge branch 'feature/add-storage' into develop
 # FIX - Conflict fix

MoonSeongKwang committed:
Merge branch 'develop' of github.com:gruut/gruut into develop

MoonSeongKwang committed:
# EDIT - storage.hpp/cpp change name
 ### 변경사항
storage가 기존 enterprise에는 leveldb 하나만으로 운영되고 있었으므로
storage라고 명명되어 있었으나, public에서는 RDB와 KV를 모두 사용되므로
storage라는 명칭은 적절하지 않아서 변경하였습니다.

기존에 연구실에서 작성된 storage_lib 파일은 json의 내용들을 읽어와서,
db_lib의 쿼리 함수들을 불러와서 각종 정보를 저장하는 역할을 맡고
있었습니다. 하지만 그 역할들은 unresolved_block_pool에서 직접적으로
수행하거나 하는 방식으로 여러 파일로 쪼개졌으므로, 해당 내용들을
이동하였습니다.
그렇게 하여 현 kv_store는 enterprise에서의 storage의 역할과 비슷하게 Key-Value Store인 levelDB만을
컨트롤하는 요소로 설계하였습니다.

MoonSeongKwang committed:
# EDIT - Eden chain's creator info delete
GP10. Genesis에 따르면, eden chain의 creator 관련 정보는 반드시 world
creator와 동일하기 때문에 생략된다고 기술되어 있습니다. 그러므로 관련
코드를 수정합니다.

MoonSeongKwang committed:
# FIX - minor

MoonSeongKwang committed:
# EDIT - FindLevelDB 수정
FindLevelDB가 정상적으로 동작하도록 FindLevelDB.cmake, CMakeList들을
수정하였습니다.

MoonSeongKwang committed:
# EDIT - DB에 관한 사항들의 위치 재조정
 ### 변경점
chain과 chain_plugin의 정확한 역할 분담을 실시하였습니다.

chain은 데이터베이스(RDB, KV-store) 전반을 컨트롤하는 요소이며,
chain_plugin은 데이터베이스에 대한 직접적인 접근은 행하지 않고, 검증과 계산 등의 더 추상적인 작업을 행합니다.
그러므로 기존에 chain_plugin에 위치했던 DBController가 chain_plugin으로 옮겨갔으며,
KV-store인 levelDB가 추가되므로 정확한 역할 구분을 위해 RdbController로
이름이 변경되었습니다. KvStore로 되어있던 leveldb의 요소 또한 통일성을
위해 KvController로 이름이 변경되었습니다.

MoonSeongKwang committed:
# ADD - KeyValue store add, World/Chain info save
levelDB를 추가하여 KV가 동작하는것을 확인하였습니다.
world_create에 의한 world와 eden 체인에 대한 정보를 저장합니다.

저장할 때의 함수 구조 등은 일단 떠오르는 가장 단순한 방법으로 구현해둬서 추후
구현에 따라 개선하여 리팩토링할 예정입니다.
이후로는 unresolved block pool에 대해 백업을 구현할 예정이며, eden 체인
외의 타 로컬체인도 추가할 수 있도록 추가 구현이 필요합니다.

MoonSeongKwang committed:
Merge branch 'develop' of github.com:gruut/gruut into develop

MoonSeongKwang committed:
# FIX - Conflict fix

morishjs commented on src/plugins/chain_plugin/CMakeLists.txt:
```diff
@@ -9,24 +9,35 @@ file(GLOB HEADERS
+
+find_package(Mysql)
+if (MYSQL_FOUND)
+    target_link_libraries(chain_plugin ${MYSQL_LIBRARIES} ${MYSQL_EMBEDDED_LIBRARIES})
```
${MYSQL_EMBEDDED_LIBRARIES} 이거는 어디에 필요한건가요? 이거 제거하고 빌드해서 돌려봐도 아무 문제가 없는걸 봐서 사용이 안되는 라이브러리 같은데...

MoonSeongKwang replied:
FindMysql쪽에 명시되어있길래 그대로 가져다 쓴건데, 굳이 필요 없다면 삭제할게요

morishjs commented on src/plugins/chain_plugin/chain_plugin.cpp:
```diff
@@ -152,8 +138,7 @@ class ChainPluginImpl {
     chain->startup(genesis_state);
-
-    rdb_controller = make_unique<DBController>(dbms, table_name, db_user_id, db_password);
+    chain->rdb_controller = make_unique<RdbController>(dbms, table_name, db_user_id, db_password);
```
chain 이 controller와 has-a 관계가 되었으면 chain_plugin에서는 controller의 존재를 모르는게 의존성 관리면에서 좋을 것 같은데요. 여기서 초기화하는 것보다 제 의견은 Chain#startup 이나 생성자가 호출될때 controller도 같이 초기화하면 될 것 같습니다.

MoonSeongKwang replied:
@park53kr 
이걸 여기서 굳이 이렇게 초기화한 이유는 `dbms, table_name, db_user_id, db_password` 때문이었습니다. db에 관한 정보를 줘야 세션을 생성해서 가지고있을 수 있어서 그런데, Chain에서 단독으로 초기화하면 저 정보들을 받아올 수가 없어서요.

이 문제를 수정하려면 어찌됐건 데이터베이스 접근 권한에 관한 정보를 chain에게 건네줘야 하는데, controller에 직접 전달하지 말고 별개의 set 함수같은걸 새로 만들까요?

morishjs replied:
@SeongKwangMoon 
따로 setter 함수 만들 필요없이 그냥 Chain 초기화할때나 startup할때 넘겨주면 될 것 같은데요

morishjs commented on src/plugins/chain_plugin/chain_plugin.cpp:
```diff
@@ -200,7 +185,7 @@ class ChainPluginImpl {
     logger::INFO("first block 0th cid: " + first_block.getTransactions()[0].getContractId());
 
-    rdb_controller->insertBlockData(first_block);
+    chain->rdb_controller->insertBlockData(first_block);
```
위와 마찬가지로 `chain->insertBlockData` 만 노출시키고 `chain->insertBlockData` -> `controller->inserBlockData` message delegation 하면 될 것 같습니다.

참조: Law of Demeter, https://en.wikipedia.org/wiki/Law_of_Demeter

park53kr commented:
@SeongKwangMoon 
1. rpc_services.cpp 에 closing curly-braces 가 하나 더 추가되어서 빌드가 안되던데, 그것도 제거해주세요.
2. leveldb 때문에 생성되는 파일들은 .gitignore에 추가해서 repo에 포함되지 않도록 해주세요. 한번 돌리니까 데이터 파일들이 생성되네요.

<img width="345" alt="image" src="https://user-images.githubusercontent.com/6479241/57668054-01672700-7641-11e9-8083-1bed13756a3c.png">

MoonSeongKwang committed:
# FIX - Requested change
 ### 변경점
1. CMakeLists의 불필요한 부분을 지웠습니다.
2. Chain의 생성자에 db 접근 관련 정보를 줌으로써 Chain 자체적인
초기화에서도 rdb_controller에 정보를 줄 수 있게 수정하였습니다.
3. rpc_services.cpp에 } 하나가 추가되어있는걸 수정하였씁니다.
4. gitignore에 /leveldb 디렉토리를 추가함으로써 git에 연동되지 않도록
하였습니다.

아직 controller에 직접 접근하는 문제는 해결하지 않았습니다.

MoonSeongKwang committed:
# EDIT - Message delegate
controller에 외부에서 직접적으로 접근하지 않고 한 단계 거쳐가도록
하였습니다.

SeongKwangMoon commented:
@park53kr 
```
void Chain::insertBlockData(gruut::Block &first_block) {
  rdb_controller->insertBlockData(first_block);
}
```
의 커밋에도 반영해뒀는데, 이런식으로 Chain에 public 함수 하나 만들어서 그대로 전달만 하도록 구현하면 되나요? 구조적으로는 말씀하신 사항의 의도가 이해는 되는데, 성능상의 이슈는 없을까 싶긴 하네요.

park53kr commented:
@SeongKwangMoon 
네 성능상에 문제 없습니다.

SeongKwangMoon commented:
그러면 아예 controller들을 private로 돌려놓겠습니다.
추후 구현에서도 이런 식으로 구현하겠습니다.

MoonSeongKwang committed:
# EDIT - Controllers changed private
 ### 변경점
1. public에 있던 RDB, KV 컨트롤러가 private로 옮겨졌습니다.
각 스토리지에 접근하기 위해서는 Chain에 있는 wrapper를 이용하여 각
컨트롤러로 전달되어 처리됩니다.

2. 각 파일들을 수정하다 보니 hpp only로 구현되어있던 block_validator가
중복 선언으로 에러가 나서 cpp로 내용 구현을 옮겼고, 원래 스토리지들을
관리하려던 storage_manager의 역할을 Chain이 모두 하고있어서 해당 파일은
삭제하였습니다.

SeongKwangMoon commented:
수정하다보니 파일들이 얽혀있어서 조금 시간이 걸렸는데, 이번 수정만 한 번 훑어봐주시면 merge해도 될 것 같습니다.

park53kr commented:
@SeongKwangMoon 
머지해도 될 것 같습니다.
